src/io/image_olm_parser.rs

```diff
@@ -36,36 +38,91 @@ pub fn render(
     let graph_width = width / chunk_size; // in chunks
     let contradiction_key = key.len();
 
-    graph
-        .vertices
+    graphs
         .into_iter()
-        .map(|labels| {
-            labels
-                .is_singleton()
-                .then(|| chunks.get_index(labels.imax()).map(|t| t.0.clone()))
-                .flatten()
-        })
-        .enumerate()
-        .for_each(|(chunk_index, opt_chunk)| {
-            let (chunk_x, chunk_y) = index_to_coords(chunk_index, graph_width);
-            // project to pixel coordinates
-            let top_left_pix_x = chunk_x * chunk_size;
-            let top_left_pix_y = chunk_y * chunk_size;
-
-            let chunk = match opt_chunk {
-                None => DMatrix::from_element(chunk_size, chunk_size, contradiction_key),
-                Some(chunk) => chunk
-            };
-
-            chunk
-                .iter()
+        .for_each(|graph| {
+            graph
+                .vertices
+                .into_iter()
+                // Vec<Multiset> => Vec<Vec<DMatrix>>
+                // map each non-zero multiset label into a corresponding DMatrix
+                .map(|vertex| {
+                    vertex
+                        .into_iter()
+                        .enumerate()
+                        .fold(Vec::new(), |mut acc, (label, frequency)| {
+                            if frequency > 0 {
+                                let (chunk, _) = chunks.get_index(label).unwrap();
+                                acc.push(chunk.clone())
+                            }
+                            acc
+                        })
+                })
+                // Vec<Vec<DMatrix>> => Vec<Option<DMatrix>>
+                // map to options depending on whether the set of chunks is empty
+                .map(|label_list| label_list.is_empty().not().then(|| label_list))
```

Comment:
The `label_list` being empty provides the same information as an optional non empty `label_list`, which is that we have chunks or not. So converting to an option here and checking for `None` later only to provide a vec of chunks is redundant. We may as well just add `DMatrix::from_element(chunk_size, chunk_size, contradiction_key)` to the `label_list` right here (if it's empty) and avoid turning into an option altogether.